### PR TITLE
PEGASUS: fix bug 11758.

### PR DESCRIPTION
**Pull request by fedor4ever**
See - https://bugs.scummvm.org/ticket/11758
PEGASUS: colossal memory leak, wan't start.

### CODE REVIEW

**Comment by sluicebox** [comment]
Could you rename this commit to conform to the guidelines? https://wiki.scummvm.org/index.php/Commit_Guidelines

The prefix should be COMMON since this doesn't touch the Pegasus engine. The summary should describe the change as opposed to just a bug number.

I'm confused if this reverts 1bb73862a10ab93ad884c9dec2908dd092cbc21e or not. Do ISOBuster files still work? That commit commented out a line and added the comment "Files produced by ISOBuster are not padded, thus, compare with the actual size", this PR restores the line but leaves that comment, and that doesn't sound right.

My interest is so that our tester who uses ISOBuster can still test my Mac stuff =)

File: common/macresman.cpp [hunk 1]
```diff
@@ -320,15 +315,10 @@ bool MacResManager::isMacBinary(SeekableReadStream &stream) {
 		//uint32 rsrcSizePad = (((rsrcSize + 127) >> 7) << 7);
 
 		// Length check
-		if (MBI_INFOHDR + dataSizePad + rsrcSize <= (uint32)stream.size()) {
-			resForkOffset = MBI_INFOHDR + dataSizePad;
-		}
+		if (MBI_INFOHDR + dataSizePad + rsrcSize == (uint32)stream.size())
```
Comment (criezy) [inline common/macresman.cpp:318]: That test is incorrect because with that change only macbinary files created with ISOBuster will (or at least might) work.

A test that might work for both ISOBuster and standard macbinary files would be:
```c++
if (MBI_INFOHDR + dataSizePad + rsrcSize == (uint32)stream.size() || MBI_INFOHDR + dataSizePad + rsrcSizePad == (uint32)stream.size())
```

File: common/macresman.cpp [hunk 2]
```diff
@@ -357,10 +347,10 @@ bool MacResManager::loadFromMacBinary(SeekableReadStream &stream) {
 
 		uint32 dataSizePad = (((dataSize + 127) >> 7) << 7);
 		// Files produced by ISOBuster are not padded, thus, compare with the actual size
-		//uint32 rsrcSizePad = (((rsrcSize + 127) >> 7) << 7);
+		uint32 rsrcSizePad = (((rsrcSize + 127) >> 7) << 7);
 
 		// Length check
-		if (MBI_INFOHDR + dataSizePad + rsrcSize <= (uint32)stream.size()) {
+		if (MBI_INFOHDR + dataSizePad + rsrcSizePad <= (uint32)stream.size()) {
```
Comment (criezy) [inline common/macresman.cpp:353]: And that change is also incorrect because it means that macbinary created by ISOBuster will fail to load.
Combined with the previous issue I pointed above it means that the commit as it currently is will likely break all macbinary files. The check here should be the same check as in `isMacBinary()`.

**Review by criezy — CHANGES_REQUESTED** [review]
It seems to me that your change will completely break reading macbinary files (created by ISOBuster or not).

The cleaning in the code is good. But the logic for the two checks in `isMacBinary()` and `loadMacBinary()` needs to be changed (and also they are currently not consistent).

**Comment by criezy** [comment]
I have also requested a review from @sev- 
Since I am not familiar with ISOBuster and he  wrote the code, it might be good for him to confirm that:
`MBI_INFOHDR + dataSizePad + rsrcSize == (uint32)stream.size()` is true for ISOBuster macbinary and thus that the following condition should work in all macbinary cases.
```c++
if (MBI_INFOHDR + dataSizePad + rsrcSize == (uint32)stream.size() || MBI_INFOHDR + dataSizePad + rsrcSizePad == (uint32)stream.size())
```

**Comment by fedor4ever** [comment]
Can you trace with debugger pegasus engine? Maybe we find better solution.

**Comment by fedor4ever** [comment]
Well, I try to write small testcase because it hard to undersatnd whats goin on.

Found suspicious code:
bool QuickTimeDecoder::loadFile(const Common::String &filename) {
	if (!Common::QuickTimeParser::parseFile(filename))
		return false;

	init();
	return true;
}

Compiles and runs but:

int main()
{
    for(auto x: data_files){
        cout << x << endl;
        initFromMovieFile(x);
    }
    return 0;
}

void initFromMovieFile(const Common::String &fileName)
{
	if (!Common::QuickTimeParser::parseFile(fileName)) { // <<< cannot call member function without object
		// Replace any colon with an underscore, since only Mac OS X
		// supports that. See PegasusEngine::detectOpeningClosingDirectory()
		// for more info.
		Common::String newName(fileName);
		if (newName.contains(':'))
			for (uint i = 0; i < newName.size(); i++)
				if (newName[i] == ':')
					newName.setChar('_', i);

		if (!Common::QuickTimeParser::parseFile(newName))
			error("Could not load video '%s'", fileName.c_str());
	}
};

**Comment by fedor4ever** [comment]
I was surprised there no tests. How this project doesn't fall apart?

**Comment by digitall** [comment]
@fedor4ever: Because people fix it... and I think you are misreading that code hence why it is failing for you in your modified version.

**Comment by digitall** [comment]
```
bool QuickTimeDecoder::loadFile(const Common::String &filename) {
	if (!Common::QuickTimeParser::parseFile(filename))
		return false;
```
looks like a call to a static method... but that is not a static class method: https://github.com/scummvm/scummvm/blob/master/common/quicktime.cpp#L62 which would be a bug... but should be one that the compiler would warn or error on... so why does that work?

Because it is not a call to a static class method, but to a superclass method... QuickTimeDecoder inherits several superclasses:
https://github.com/scummvm/scummvm/blob/master/video/qt_decoder.h#L61

The critical one here is QuickTimeAudioDecoder which inherits from QuickTimeParser hence why calling that method from the superclass is valid:
https://github.com/scummvm/scummvm/blob/master/audio/decoders/quicktime_intern.h#L48

**Comment by digitall** [comment]
Looking at the original code which calls this, it should be fine assuming I am correct in my C++ reading:
https://github.com/scummvm/scummvm/blob/master/engines/pegasus/movie.cpp#L62

**Comment by fedor4ever** [comment]
> Looking at the original code which calls this, it should be fine assuming I am correct in my C++ reading:
> https://github.com/scummvm/scummvm/blob/master/engines/pegasus/movie.cpp#L62

Yes, you correct. I read as call to static method and was wrong.

**Comment by digitall** [comment]
@fedor4ever : No problem. I thought the same thing when I first read it... but no compiler warning / error made me suspect I was missing something.

File: common/macresman.h [hunk 3]
```diff
@@ -72,20 +72,11 @@ class MacResManager {
...
-	/**
-	 * Open a Mac data/resource fork pair from within the given archive.
-	 *
-	 * @param path The path that holds the forks
-	 * @param fileName The base file name of the file
-	 * @note This will check for the raw resource fork, MacBinary, and AppleDouble formats.
-	 * @return True on success
-	 */
-	bool open(const String &fileName, Archive &archive);
+	bool open(const String &fileName, Archive &archive = SearchMan);
```
Comment (sev-) [inline common/macresman.h:79]: What is the purpose of this change?

File: common/macresman.cpp (hunk 1 above)
Comment (sev-) [reply common/macresman.cpp:318]: Indeed, this just breaks things. What is the reasoning behind this change?

**Comment by sev-** [comment]
This PR is a whole dirty hack without proper testing performed and no understanding of the code.

Closing
